datasets/flwr_datasets/partitioner/semantic_partitioner.py

```diff
@@ -0,0 +1,493 @@
+            for i in partitions:
+                partition_id_to_indices[clusters[label][i]].extend(  # type: ignore
+                    label_cluster_list[label][i]  # type: ignore
+                )
+
+        # Shuffle the indices not to have the datasets with targets in sequences like
+        # [00000, 11111, ...]) if the shuffle is True
+        if self._shuffle:
+            for indices in partition_id_to_indices.values():
+                # In place shuffling
+                self._rng_numpy.shuffle(indices)
+        self._partition_id_to_indices = partition_id_to_indices
+        self._partition_id_to_indices_determined = True
+
+    def _preprocess_dataset_images(self) -> NDArrayFloat:
+        images = np.array(self.dataset[self._data_column_name], dtype=float)
+        if len(images.shape) == 3:  # [B, H, W]
+            images = np.reshape(
+                images, (images.shape[0], 1, images.shape[1], images.shape[2])
+            )
+        elif len(images.shape) == 4:  # 2D
+            # [H, W, C]
+            if images.shape[3] < min(images.shape[1], images.shape[2]):
+                images = np.transpose(images, (0, 3, 1, 2))
+            # [C, H, W]
+            elif images.shape[1] < min(images.shape[2], images.shape[3]):
+                pass
+        else:
+            raise ValueError(f"The image shape is not supported. Now: {images.shape}")
+        return images
+
+    def _check_num_partitions_correctness_if_needed(self) -> None:
+        """Test whether the number of partitions is valid."""
+        if not self._partition_id_to_indices_determined:
+            if self._num_partitions > self.dataset.num_rows:
+                raise ValueError(
+                    "The number of partitions needs to be smaller than the number of "
+                    "samples in the dataset."
+                )
+
+    def _check_pca_components_validation_if_needed(self) -> None:
+        """Test whether pca_components is in the valid range."""
+        if not self._partition_id_to_indices_determined:
+            if self._pca_components > min(
+                self.dataset.num_rows, self._efficient_output_size
+            ):
+                raise ValueError(
+                    "The pca_components needs to be smaller than "
+                    f"min(the number of samples = {self.dataset.num_rows}, "
+                    "efficient net output size = 1280) "
+                    "in the dataset or the output size of the efficient net. "
+                    f"Now: {self._pca_components}."
+                )
+
+    def _check_data_type_if_needed(self) -> None:
+        """Test whether data is image-like."""
+        if not self._partition_id_to_indices_determined:
+            features_dict = self.dataset.features.to_dict()
+            self._data_column_name = list(features_dict.keys())[0]
+            try:
+                data = np.array(
+                    self.dataset[self._data_column_name][0], dtype=np.float32
```

Comment:
This part can be replaced with checking if the column type is `datasets.Image`, if not (for reasons I can't see right now) we should index first based on the row id then the column name. The current indexing pulls all the images to memory first (since it's indexed first by a column first).

Reply:
Agree with most of it. But I think the column type should not be strict. 
Since maybe some users want to use their private datasets by `Dataset.from_dict()` or other ways and the feature dict of that dataset maybe not set properly. 
So I think in this place, we can throw a warning says the process maybe failed due to the column type of `image_column_name` is not `datasets.Image`.